docs: Add SDK Search reference (#1065)
vtex/faststore: docs/docs/reference/sdk/search/SearchProvider.mdx

import PropsComponent from "@site/src/components/PropsComponent/PropsComponent";

# SearchProvider

SearchContext provider for Faceted Search

## Import

```tsx
import { SearchProvider } from '@faststore/sdk'
```

## Usage

```tsx
<SearchProvider
  onChange={(url: URL) => window.location.href = url.href}
  itemsPerPage={2}
  sort="score_desc"
  term={null}
  page={0}
  base=""
  selectedFacets={}
>
  {children}
</SearchProvider>

> ```

Author: what about now?

## Props

<PropsComponent
    name="onChange"
    type="(url: URL) => void"
    description="Callback function when the search state is changed"
/>

<PropsComponent
    name="itemsPerPage"
    type="number"
    description="Number of items in a given page"
/>

<PropsComponent
    name="sort"
    type="SearchSort"
    description="Ordering of search results"
/>

<PropsComponent
    name="term"
    type="string | null"
    description="Full text term used on the search"
/>

<PropsComponent
    name="page"
    type="number"
    description="Current page index of search pagination"
/>

<PropsComponent
    name="base"
    type="string"
    description="Base path of the search page. Usefull when dealing with localization and prefixing paths by the locale, e.g.: /pt-br/"
/>

<PropsComponent
    name="selectedFacets"
    type="Array<{ key: string, value: string }>"

>     description="Array of selected facets on the search. Make sure NOT to include personalization facets, like channel or price tables"

Author: Sao o sales channel ou price table. Se incluir eles ai, eles vao acabar na URL. Se voce mandar essa url para outra pessoa, pode dar um erro pois essa outra pessoa nao necessariamente vai estar no mesmo sales channel de voce e vai ver outra coisa.

>     description="Array of selected facets on the search. Make sure NOT to include personalization facets, like channel or price tables"

Reviewer: Mas acho que isso é normal. O site da amazon é assim, por exemplo: buscar por X
- uma conta com prime e frete gratis do prime, o produto X1 fica com menor preço, então ele é o escolhido;
- uma conta sem prime e produto com frete, o produto X2 fica com menor preço, então ele é o escolhido;

Se o usuário 1 compartilhar o link pra o usuário 2, o usuário dois, vai ver X2 e não X1.

>     description="Array of selected facets on the search. Make sure NOT to include personalization facets, like channel or price tables"

Author: sim, mas isso vai dar todo o tipo de bug na nossa implementacao. Vai ter duas fontes de verdade de sales channel etc e cada componente vai mostrar uma coisa pois cada um vai consumir esse dado de um provedor de contexto diferente, entao é melhor nao fazermos essa feature pelo menos por enquanto

      onChange={(url: URL) => window.location.href = url.href}
      itemsPerPage={12}
      {...params}
    >
      {children}
    </SearchProvider>
  )
}
```

## Props

<PropsComponent
    name="onChange"
    type="(url: URL) => void"
    description="Callback function responsible for handling changes in the facet state."
/>

<PropsComponent
    name="itemsPerPage"
    type="number"
    description="Number of product items displayed on a given page."
/>

<PropsComponent
    name="sort"
    type="SearchSort"
    description="Sorting criteria of the search result (e.g., price_asc, orders_desc, discount_desc, etc.)."
/>

<PropsComponent
    name="term"
    type="string | null"
    description="Full-text term used on the search."
/>

<PropsComponent
    name="page"
    type="number"
    description="Current page index of search pagination."
/>

<PropsComponent
    name="base"
    type="string"
    description="Base URL path of the search page. Useful when dealing with localization and prefixing paths by locale (e.g., /pt-br/)."
/>

<PropsComponent
    name="selectedFacets"
    type="Array<{ key: string, value: string }>"
    description="Array of selected facets on the search."
/>

> :::caution WIP
> Make sure NOT to include personalization facets, like channel or price tables on selectedFacets. If you do, these facets, alongside their values, will be serialized in the url. This means that a user may end up having two channels, one from the Session context and another from the url, causing all sorts of weird bugs

Reviewer 2: ```suggestion
:::caution
Don't include personalization facets, such as sales channel and price tables, on the `selectedFacets` property. If you do so, users may end up having two channels: one from the Session context and another from the URL, which may cause all sorts of unexpected bugs.
```
